<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Skyline Models with heterochronous data</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="https://revbayes.github.io/revscripter/">RevScripter</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Skyline Models with heterochronous data</h1>
	<h3 class="subtitle">Estimating Demographic Histories with Skyline Models using heterochronous data</h3>
	<h4 class="authors">Sebastian HÃ¶hna and Ronja Billenstein</h4>
  <h5>Last modified on February 14, 2022</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul>
          <li>None</li>
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Other</strong>
        <ul id="other_files">
        
        
        
          <li><a href="/tutorials/coalescent/data/horses_heterochronous_ages.tsv">horses_heterochronous_ages.tsv</a></li>
        
          <li><a href="/tutorials/coalescent/data/horses_heterochronous_sequences.fasta">horses_heterochronous_sequences.fasta</a></li>
        
          <li><a href="/tutorials/coalescent/scripts/mcmc_heterochronous_skyline.Rev">mcmc_heterochronous_skyline.Rev</a></li>
        
        </ul>
    
</blockquote>


</div>
<h2 class="section" id="overview">Overview</h2>
<hr class="section" />

<p>This tutorial describes how to run a Skyline Analysis with heterochronous data in <code class="language-plaintext highlighter-rouge">RevBayes</code>.</p>

<h2 class="section" id="inference-example">Inference Example</h2>
<hr class="section" />

<blockquote class="info">
  <h2 id="for-your-info">For your info</h2>
  <p>The entire process of the skyline estimation can be executed by using the <strong>mcmc_skyline.Rev</strong> script in the <strong>scripts</strong> folder.
You can type the following command into <code class="language-plaintext highlighter-rouge">RevBayes</code>:</p>
  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>source("scripts/mcmc_skyline.Rev")
</code></pre></div>  </div>
  <p>We will walk you through every single step in the following section.</p>
</blockquote>

<h3 class="subsection" id="read-the-data">Read the data</h3>
<hr class="subsection" />

<!--- Start by reading in the taxa names and age information (**bears_taxa.tsv**) and the sequences (**.nex**). --->
<p>Start by reading in the aligned sequences.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sequences &lt;- readDiscreteCharacterData("data/modern_narwhals.fasta")
</code></pre></div></div>

<p>You will also need the names of the taxa and their number.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>taxa &lt;- sequences.taxa()
n_taxa &lt;- taxa.size()
</code></pre></div></div>

<h3 class="subsection" id="the-skyline-model">The Skyline Model</h3>
<hr class="subsection" />

<p>For the skyline model, you need to decide on the number of intervals.
In order to get five coalescent events per interval, we divide the number of taxa by five.
With $n$ taxa, we expect $(n-1)$ coalescent events, until there is only one lineage left.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>NUM_INTERVALS = ceil(n_taxa / 5)
</code></pre></div></div>

<p>For each interval, a population size will be estimated. Chose a prior and add a move for each population size.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>for (i in 1:NUM_INTERVALS) {
    pop_size[i] ~ dnUniform(0,1E6)
    moves.append( mvScale(pop_size[i], lambda=0.1, tune=true, weight=2.0) )
}
</code></pre></div></div>

<!--- Also, the lengths of the intervals, more specifically the points of change between intervals, have to be set.

~~~
for (i in 1:(NUM_INTERVALS-1)) {
    changePoints[i] <- i * 0.5
}
~~~
--->

<h3 class="subsection" id="the-tree">The Tree</h3>
<hr class="subsection" />

<p>Now, the will instantiate the stochastic node for the tree. The Skyline distribution function <code class="language-plaintext highlighter-rouge">dnCoalescentSkyline</code> takes the vector of population sizes the taxa as input. <!--- **(need to check whether it can take more)** --->
By chosing <code class="language-plaintext highlighter-rouge">methods="events"</code>, the interval lengths will be chosen based on the number of events.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>psi ~ dnCoalescentSkyline(theta=pop_size, method="events", taxa=taxa)
</code></pre></div></div>

<p>In order to be able to later plot and analyze the population size curve, we need to retrieve the resulting interval times.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>interval_times := psi.getIntervalAges()
</code></pre></div></div>

<p>For this analysis, we calibrate the tree based on the root age.
We chose a LogNormal distribution with a mean of $0.11$ and a standard deviation of $0.3$.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>root_age := psi.rootAge()
obs_root_age ~ dnLognormal( ln(root_age), sd=0.3 )
obs_root_age.clamp(0.11)
</code></pre></div></div>

<p>We should also add moves for the tree.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>moves.append( mvNarrow(psi, weight=5.0) )
moves.append( mvNNI(psi, weight=1.0) )
moves.append( mvFNPR(psi, weight=3.0) )
moves.append( mvSubtreeScale(psi, weight=3.0) )
moves.append( mvNodeTimeSlideUniform(psi, weight=15.0) )
moves.append( mvRootTimeScaleBactrian(psi, weight=3.0) )
moves.append( mvTreeScale(psi, weight=3.0) )
</code></pre></div></div>

<h3 class="subsection" id="substitution-model-and-other-parameters">Substitution Model and other parameters</h3>
<hr class="subsection" />

<p>Finally, sequence data should be added to the analysis. Here, we assume a GTR+$\Gamma$+I substitution model, but you can of course use others. Have a look at the <a href="/tutorials/ctmc/">Nucleotide substitution models</a> tutorial to see how you can define different substitution models.
<!--- The `dnPhyloCTMC` function can take several arguments, here it is enough to specify the tree, the Q matrix and the type of data. ---></p>

<p>For the GTR model, we need to add exchangeability rates (<code class="language-plaintext highlighter-rouge">er</code>) and stationary frequences (<code class="language-plaintext highlighter-rouge">pi</code>).
Of course, we also add moves for these.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>er_prior &lt;- v(1,1,1,1,1,1)
pi_prior &lt;- v(1,1,1,1)
er ~ dnDirichlet(er_prior)
pi ~ dnDirichlet(pi_prior)

moves.append( mvBetaSimplex(er, weight=3) )
moves.append( mvDirichletSimplex(er, weight=1) )
moves.append( mvBetaSimplex(pi, weight=2) )
moves.append( mvDirichletSimplex(pi, weight=1) )
</code></pre></div></div>
<p>This is everything needed for the Q matrix of the GTR model.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Q := fnGTR(er,pi)
</code></pre></div></div>
<p>For the $\Gamma$ extension to the GTR model, we need to draw the site rates (<code class="language-plaintext highlighter-rouge">sr</code>) from a discretized Gamma function with two parameters.
Here, we use <code class="language-plaintext highlighter-rouge">alpha</code> for both parameters.
We also need to add a move for <code class="language-plaintext highlighter-rouge">alpha</code>.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>alpha ~ dnUniform( 0.0, 1E6 )
alpha.setValue( 1.0 )
sr := fnDiscretizeGamma( alpha, alpha, 4 )
moves.append( mvScale(alpha, weight=2.0) )
</code></pre></div></div>
<p>We draw the proportion of invariant sites (<code class="language-plaintext highlighter-rouge">p_inv</code>) from a Beta distribution and add a move.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>p_inv ~ dnBeta(1,1)
moves.append( mvSlide(p_inv) )
</code></pre></div></div>
<p>The last step is to define the clock rate to be drwan from a Uniform distribution.
<!--- We set the starting value according to the original analysis published in <a class="citation" href="#"></a> and add a move. ---></p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>clock ~ dnUniform(0.0, 1.0)
moves.append( mvScale(clock, weight=2.0) )
</code></pre></div></div>
<p>The final <code class="language-plaintext highlighter-rouge">dnPhyloCTMC</code> function combines all of the previous defined parameters.
We also need to clamp the sequence data.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>seq ~ dnPhyloCTMC(tree=psi, Q=Q, siteRates=sr, pInv=p_inv, type="DNA", branchRates=clock)
seq.clamp(sequences)
</code></pre></div></div>

<h3 class="subsection" id="finalize-and-run-the-analysis">Finalize and run the analysis</h3>
<hr class="subsection" />

<p>In the end, we need to wrap our model.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymodel = model(psi)
</code></pre></div></div>

<p>Finally, we can add some monitors and then run the MCMC.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>monitors.append( mnModel(filename="output/example_skyline.log",printgen=THINNING) )
monitors.append( mnFile(filename="output/example_skyline.trees",psi,printgen=THINNING) )
monitors.append( mnFile(filename="output/example_skyline_NEs.log",pop_size,printgen=THINNING) )
monitors.append( mnFile(filename="output/example_skyline_times.log",interval_times,printgen=THINNING) )
monitors.append( mnScreen(pop_size, root_age, printgen=100) )

mymcmc = mcmc(mymodel, monitors, moves)
mymcmc.burnin(NUM_MCMC_ITERATIONS*0.1,100)
mymcmc.run(NUM_MCMC_ITERATIONS, tuning = 100)
</code></pre></div></div>

<h2 class="section" id="results">Results</h2>
<hr class="section" />

<p>After running your analysis, you can plot the results using the <code class="language-plaintext highlighter-rouge">R</code> package <code class="language-plaintext highlighter-rouge">RevGadgets</code>.</p>

<figure id="example_skyline"><p><img src="figures/example_skyline_scaled.png" width="800" /></p>
<figcaption>This is how the resulting skyline plot should roughly look like.</figcaption>
</figure>

<ol class="bibliography"></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

  </body>
</html>
