<!doctype html>
<html lang="en">
<link rel="icon" type="image/png" href="/assets/img/favicon.png" >
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="search-domain" value="https://revbayes.github.io/">
    <link href="https://fonts.googleapis.com/css?family=Raleway" rel="stylesheet">
    <link rel="stylesheet" href="/assets/css/syntax.css">
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="/assets/css/main.css" />
    <title>RevBayes: Coalescent Analyses</title>
  </head>
  <body>
    <div class="container">
      <nav class="navbar navbar-default navbar-fixed-top">
  <div class="container-fluid">
    <div class="navbar-header">
      <a href="/" class="pull-left">
        
        <img class="navbar-logo" src="/assets/img/aquabayes-desaturated.png" alt="RevBayes Home" />
        
      </a>

      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar" align="right"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>

    </div>
    <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
      <ul class="nav navbar-nav">
        <li><a href="/download">Download</a></li>
        <li><a href="/tutorials/">Tutorials</a></li>
        <li><a href="/documentation/">Documentation</a></li>
        <li><a href="https://revbayes.github.io/revscripter/">RevScripter</a></li>
        <li><a href="/workshops/">Workshops</a></li>
        <li><a href="/jobs/">Jobs</a></li>
        <li><a href="/developer/">Developer</a></li>
        <li><<a href="/gui">GUI</a></li>
      </ul>
      <!-- <form class="navbar-form navbar-right" role="search" id="search" onsubmit="google_search(); return false;">
        <div class="form-group">
          <input type="text" id="google-search" placeholder="Search..." aria-label="Google site search">
        </div>
      </form> -->
    </div>
  </div>
</nav>

      <div class="titlebar">
	<h1 class="maintitle">Coalescent Analyses</h1>
	<h3 class="subtitle">Estimating Demographic Histories with Coalescent Processes</h3>
	<h4 class="authors">Ronja Billenstein and Sebastian Höhna</h4>
  <h5>Last modified on February 23, 2022</h5>
</div>


<div class="sidebar no-print">
<blockquote class="overview" id="overview">
  <h2>Overview</h2>
  
  <div class="row">
    <div class="col-md-9">
        <strong>Prerequisites</strong>
        
          <ul id="prerequisites">
          
            <li><a href="/tutorials/intro/">Getting Started with RevBayes and Rev Language Syntax</a></li>
          
            <li><a href="/tutorials/mcmc/">Introduction to Markov chain Monte Carlo (MCMC) Sampling</a></li>
          
            <li><a href="/tutorials/ctmc/">Nucleotide substitution models</a></li>
          
          </ul>
        
    </div>
  </div>
  
</blockquote>





<blockquote class="tutorial_files" id="tutorial_files">
    <h2>Data files and scripts</h2>
    
        
        <strong>Other</strong>
        <ul id="other_files">
        
        
        
          <li><a href="/tutorials/coalescent/data/horses_heterochronous_ages.tsv">horses_heterochronous_ages.tsv</a></li>
        
          <li><a href="/tutorials/coalescent/data/horses_heterochronous_sequences.fasta">horses_heterochronous_sequences.fasta</a></li>
        
          <li><a href="/tutorials/coalescent/data/horses_homochronous_sequences.fasta">horses_homochronous_sequences.fasta</a></li>
        
        </ul>
    
</blockquote>


</div>
<blockquote class="info">
  <h2 id="for-your-info">For your info</h2>
  <p>This tutorial and the included exercises are currently under construction.</p>
</blockquote>

<h2 class="section" id="overview">Overview</h2>
<hr class="section" />

<p>This tutorial describes how to run a demographic analysis using the coalescent process in <code class="language-plaintext highlighter-rouge">RevBayes</code>.
Demographic inference is about estimating population dynamics and in this tutorial, we will specifically focus on population size estimation.
The coalescent process provides a flexible way of estimating population size trajectories through time.
The input data are usually sequence alignments or estimated trees.
In case of sequence data, trees and population size parameters will be jointly estimated.
Here, we consider different types of analysis, starting with a constant demographic history and doing more complex analyses in later exercises.
The first analyses will consider data from homochronous samples, <em>i.e.</em>, data from samples that have all been collected at the same time.
In the second part of the exercises, you will be asked to also analyze data from heterochronous samples, <em>i.e.</em>, samples that have been collected at different points in time.
The data for these analyes are taken from <a class="citation" href="#Vershinina2021">Vershinina et al. (2021)</a>.</p>

<!--- ### Why?! --->

<!---
<h3 class="subsection" id="the-coalescent">The Coalescent</h3><hr class="subsection">
The coalescent process is constructing a tree backwards in time.
Starting from the samples, lineages are merged (*i.e.* coalesced), always two at a time.
Under the coalescent process, the waiting time between two coalescent events is exponentially distributed and depends on the number of 'active' lineages and the effective population size $N_e$.
Active lineages are the ones that can coalesce, the number is reduced by one with every coalescent event.
The coalescent process was first introduced by Kingman in 1982 for a constant population size <a class="citation" href="#Kingman1982">(Kingman 1982)</a>.
Griffiths and Tavaré then extended the model to be applicable to varying population sizes <a class="citation" href="#Griffiths1994">(Griffiths et al. 1994)</a>.
--->
<!--- Also, samples from different ages can be included <a class="citation" href="#"></a>. (Should go to heterochronous part) --->

<!--- ### Add figure!! --->

<h2 class="section" id="preparation">Preparation</h2>
<hr class="section" />

<blockquote class="instruction">
  <p>In order to perform the different analyses in this tutorial, you will need to create a directory on your computer for this tutorial and download a few files.</p>
</blockquote>

<h3 class="subsection" id="the-data">The Data</h3>
<hr class="subsection" />

<blockquote class="instruction">
  <p>In the tutorial directory on your computer, create a subdirectory called “data” and download the data files that you can find on the left of this page.</p>
</blockquote>

<p>You should now have the following files in your data folder:</p>

<ul>
  <li>
    <p>horses_homochronous_sequences.fasta: an alignment in FASTA format, containing sequences of 36 horse taxa, all sampled at the same time.</p>
  </li>
  <li>
    <p>horses_heterochronous_sequences.fasta: an alignment in FASTA format, containing sequences of 173 horse taxa, sampled at various times.</p>
  </li>
  <li>
    <p>horses_heterochronous_ages.tsv: a tab seperated table listing the heterochronous horse samples and their ages. For extant taxa, the minimum age is 0.0 (i.e. the present).</p>
  </li>
</ul>

<h3 class="subsection" id="the-scripts">The Scripts</h3>
<hr class="subsection" />

<p>It is useful to create <code class="language-plaintext highlighter-rouge">.Rev</code> scripts for the different analyes.
We will also provide a full script in every tutorial that you can easily run for the whole analysis.</p>

<blockquote class="instruction">
  <p>Please create a “scripts” directory in the tutorial directory.</p>
</blockquote>

<h2 class="section" id="exercises">Exercises</h2>
<hr class="section" />

<blockquote class="instruction">
  <p>You can begin by clicking on the first exercise!</p>
</blockquote>

<p>If you want to turn back to this page, it is listed on the left as prerequisite in all the exercises.
The exercise for heterochronous data is based on the exercises for homochronous data.
It aims at highlighting the changes you have to make when considering samples with different ages.
It is therefore recommended to do the exercises for homochronous data first.</p>

<p>The first five exercises work with homochronous data, the last one with heterochronous data:</p>
<ol>
  <li><a href="/tutorials/coalescent/constant">The constant coalescent model</a></li>
  <li><a href="/tutorials/coalescent/skyline">The skyline model</a></li>
  <li><a href="/tutorials/coalescent/GMRF">The Gaussian Markov Random Field (GMRF) model</a></li>
  <li><a href="/tutorials/coalescent/GMRF_treebased">The GMRF model with trees as input data</a></li>
  <li><a href="/tutorials/coalescent/piecewise">A piecewise model</a></li>
</ol>

<!--- 6. [Heterochronous data](/tutorials/coalescent/heterochronous) --->

<h2 class="section" id="summary">Summary</h2>
<hr class="section" />

<p>After doing all the exercises, you can compare the resulting population sizes.
Have a look at the <a href="/tutorials/coalescent/summary">summary</a>.</p>

<ol class="bibliography"><li><span id="Griffiths1994">Griffiths R.C., Tavare S., Bodmer W.F., Donnelly P.J. 1994. Sampling theory for neutral alleles in a varying environment. Philosophical Transactions of the Royal Society of London. Series B: Biological Sciences. 344:403–410.</span>

<a href="https://doi.org/10.1098/rstb.1994.0079">10.1098/rstb.1994.0079</a>

</li>
<li><span id="Kingman1982">Kingman J.F.C. 1982. On the Genealogy of Large Populations. Journal of Applied Probability. 19:27–43.</span>

<a href="https://doi.org/10.2307/3213548">10.2307/3213548</a>

</li>
<li><span id="Vershinina2021">Vershinina A.O., Heintzman P.D., Froese D.G., Zazula G., Cassatt-Johnstone M., Dalén L., Der Sarkissian C., Dunn S.G., Ermini L., Gamba C., Groves P., Kapp J.D., Mann D.H., Seguin-Orlando A., Southon J., Stiller M., Wooller M.J., Baryshnikov G., Gimranov D., Scott E., Hall E., Hewitson S., Kirillova I., Kosintsev P., Shidlovsky F., Tong H.-W., Tiunov M.P., Vartanyan S., Orlando L., Corbett-Detig R., MacPhee R.D., Shapiro B. 2021. Ancient horse genomes reveal the timing and extent of dispersals across the Bering Land Bridge. Molecular Ecology. 30:6144–6161.</span>

<a href="https://doi.org/https://doi.org/10.1111/mec.15977">https://doi.org/10.1111/mec.15977</a>

</li></ol>

<script type="text/javascript">
var _ol = document.querySelectorAll('ol');
for (var i = 0, elem_ol; elem_ol = _ol[i]; i++) {
	if ( elem_ol.classList == "bibliography" ) {
		var _li = elem_ol.getElementsByTagName("li");
		//for (var j = 0, elem_li; elem_li = _li[j]; j++)
		//{
		//	elem_li.innerHTML = elem_li.innerHTML.replace(/(https?:\/\/)([^\s<]+)/,"<a href=\"$1$2\">$2");
		//}
		if(_li.length > 0)
			elem_ol.outerHTML = "<h2 class='references'>References</h2><hr class='references'>"+elem_ol.outerHTML
	}
}
</script>

      <br>
<footer>
  <div class="container">
  <div class="row">
    <div class="col-sm-12" align="center">
      <a href="https://github.com/revbayes">GitHub</a> | <a href="/license">License</a> | <a href="/citation">Citation</a> | <a href="https://groups.google.com/forum/#!forum/revbayes-users">Users Forum</a>
    </div>
  </div>
  <br>
  </div>
</footer>

    </div>
    <script src="/assets/js/vendor/jquery.min.js"></script>
<script src="/assets/js/vendor/FileSaver.min.js"></script>
<script src="/assets/js/vendor/jszip.min.js"></script>
<script src="/assets/js/vendor/bootstrap.min.js"></script>

<script type="text/javascript">
// Add default language
$(":not(code).highlighter-rouge").each(function() {
  
  if( this.classList == "highlighter-rouge") {
    this.classList = "Rev highlighter-rouge";
  }
  
});
// $("code.highlighter-rouge").each(function() {
//   
//   if( this.classList == "highlighter-rouge") {
//       this.classList = "Rev highlighter-rouge";
//   }
//   
// });
</script>
<script type="text/javascript"
    src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
    MathJax.Hub.Config({
    tex2jax: {
      inlineMath: [['$','$'], ['\\(','\\)']],
      processEscapes: true
    }
    });
    MathJax.Hub.Queue(function () {
      $(".aside").each(function() {
          $("div .MathJax", this).hide();
      });
    });
</script>
<script src="/assets/js/base.js"></script>

  </body>
</html>
